jans-cedarling/cedarling/src/jwt/decoding_strategy.rs

```diff
@@ -5,123 +5,184 @@
-/// appropriate strategy can be selected based on the bootstrap config.
+/// This enum defines two strategies for decoding JWT tokens: `WithoutValidation`
+/// for decoding without validation and `WithValidation` for decoding with validation
+/// using a key service and supported algorithms.
 pub enum DecodingStrategy {
-    /// Decoding strategy that does not perform validation.
+    /// Decoding strategy that skips all validation.
     WithoutValidation,
 
     /// Decoding strategy that performs validation using a key service and supported algorithms.
     WithValidation {
-        key_service: Arc<dyn GetKey>,
+        key_service: KeyService,
         supported_algs: Vec<jwt::Algorithm>,
     },
 }
 
 impl DecodingStrategy {
-    /// Creates a new decoding strategy that skips validation.
-    ///
-    /// This method initializes the `WithoutValidation` strategy, which
-    /// decodes JWT tokens without any validation checks
+    /// Creates a new decoding strategy that does not perform validation.
     pub fn new_without_validation() -> Self {
         Self::WithoutValidation
     }
-}
 
-/// Trait implementation for decoding a JWT token.
-impl Decode for DecodingStrategy {
-    // Decode a JWT according to the current decoding strategy.
-    fn decode<T: DeserializeOwned>(
+    /// Creates a new decoding strategy that performs validation.
+    ///
+    /// This strategy uses the provided dependency map to configure a key service
+    /// and validates tokens based on the specified algorithms.
+    ///
+    /// # Errors
+    /// Returns an error if the specified algorithm is unrecognized or the key service initialization fails.
+    pub fn new_with_validation(
+        config_algs: Vec<jwt::Algorithm>,
+        trusted_idps: Vec<TrustedIssuer>,
+    ) -> Result<Self, Error> {
+        // initialize the key service with OpenID configuration endpoints
+        let openid_conf_endpoints = trusted_idps
+            .iter()
+            .map(|x| x.openid_configuration_endpoint.as_ref())
+            .collect();
+        let key_service = KeyService::new(openid_conf_endpoints).map_err(Error::KeyService)?;
+
+        Ok(Self::WithValidation {
+            key_service,
+            supported_algs: config_algs,
+        })
+    }
+
+    /// Decodes a JWT token according to the current decoding strategy.
+    ///
+    /// # Arguments
+    /// - `jwt` - The JWT string to decode.
+    /// - `iss` - Optional expected issuer for validation.
+    /// - `aud` - Optional expected audience for validation.
+    /// - `sub` - Optional expected subject for validation.
+    /// - `validate_nbf`: A boolean indicating whether to validate the "not before" claim.
+    /// - `validate_exp`: A boolean indicating whether to validate the expiration claim.
+    ///
+    /// # Errors
+    /// Returns an error if decoding or validation fails.
+    pub fn decode<T: DeserializeOwned>(
         &self,
         jwt: &str,
         iss: Option<impl ToString>,
```

Comment:
For these `Option<impl ToString>, what will be passed in to this function other than `Option<String>`?

Reply:
setting up validation using the `jsonwebtoken` create is kinda messy since it's forcing me to pass in variables in a certain type... 

to implement validation, using the `jsonwebtoken` crate, i need to do something like this:
```rust
// set up validation rules
let mut validator = jwt::Validation::new(header.alg);
validator.required_spec_claims.clear();
validator.validate_nbf = args.validate_nbf;
validator.validate_exp = args.validate_exp;
if let Some(iss) = args.iss {
    validator.set_issuer(&[iss]);
} else {
    validator.iss = None;
}
if let Some(aud) = args.aud {
    validator.set_audience(&[aud]);
} else {
    validator.validate_aud = false;
}
validator.sub = args.sub.map(|sub| sub.to_string());
```
here, the `iss` field is asking for a `&[]` but for our use-case, it mostly will either be `None` to indicate "do not perform validation on the `iss` field" or `Some(impl ToString)` which uses whatever inside the `Some` to implement the validation. It's kinda messy right now and I am thinking we can implement our own validation since it's not that complicated and we can define our own signature.

... though i don't know if i understood your question correctly -- we're you suggesting we can use something other than `Option<impl ToString>`?